lib/Dialect/Seq/Transforms/HWMemSimImpl.cpp

```diff
@@ -700,6 +703,39 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
   });
 }
 
+struct PrefixingInliner : public mlir::InlinerInterface {
+  StringRef prefix;
+  PrefixingInliner(MLIRContext *context, StringRef prefix)
+      : mlir::InlinerInterface(context), prefix(prefix) {}
+
+  bool isLegalToInline(Region *dest, Region *src, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    return true;
+  }
+  bool isLegalToInline(Operation *op, Region *dest, bool wouldBeCloned,
+                       IRMapping &valueMapping) const override {
+    return true;
+  }
+  void handleTerminator(Operation *op,
+                        ArrayRef<Value> valuesToRepl) const override {
+    assert(isa<hw::OutputOp>(op));
+    for (auto [from, to] : llvm::zip(valuesToRepl, op->getOperands()))
+      from.replaceAllUsesWith(to);
+  }
+
+  void processInlinedBlocks(
+      iterator_range<Region::iterator> inlinedBlocks) override {
+    for (Block &block : inlinedBlocks)
+      block.walk([&](Operation *op) {
+        if (auto name = op->getAttrOfType<StringAttr>("name");
+            name && !name.getValue().empty()) {
+          op->setAttr("name", StringAttr::get(name.getContext(),
+                                              prefix + "_" + name.getValue()));
+        }
+      });
+  }
+};
```

Comment:
Placing this in an anonymous `namespace` would hide it from external users. This is annoyingly dangerous as it doesn't handle inner symbols. It would be fine to not handle these and to instead not inline memories which have them.

Consider the following:

``` mlir
module {
  sv.macro.decl @RANDOM
  // This is a hierarchical path op which references inner symbol `@Foo::@mem`, i.e., the original instance.
  hw.hierpath @Bar [@Foo::@mem]
  hw.generator.schema @FIRRTLMem, "FIRRTL_Memory", ["depth", "numReadPorts", "numWritePorts", "numReadWritePorts", "readLatency", "writeLatency", "width", "maskGran", "readUnderWrite", "writeUnderWrite", "writeClockIDs", "initFilename", "initIsBinary", "initIsInline"]
  hw.module.generated @memory_8x32, @FIRRTLMem(in %R0_addr : i3, in %R0_en : i1, in %R0_clk : i1, out R0_data : i32, in %W0_addr : i3, in %W0_en : i1, in %W0_clk : i1, in %W0_data : i32) attributes {depth = 8 : i64, initFilename = "", initIsBinary = false, initIsInline = false, maskGran = 32 : ui32, numReadPorts = 1 : ui32, numReadWritePorts = 0 : ui32, numWritePorts = 1 : ui32, readLatency = 1 : ui32, readUnderWrite = 0 : i32, width = 32 : ui32, writeClockIDs = [0 : i32], writeLatency = 1 : ui32, writeUnderWrite = 1 : i32}
  hw.module @Foo(in %r_addr : i3, in %r_en : i1, in %r_clk : i1, out r_data : i32, in %w_addr : i3, in %w_en : i1, in %w_clk : i1, in %w_data : i32, in %w_mask : i1) {
    // The inner symbol `@Foo::@mem` is defined here:
    %memory_ext.R0_data = hw.instance "memory_ext" sym @mem @memory_8x32(R0_addr: %r_addr: i3, R0_en: %r_en: i1, R0_clk: %r_clk: i1, W0_addr: %w_addr: i3, W0_en: %0: i1, W0_clk: %w_clk: i1, W0_data: %w_data: i32) -> (R0_data: i32)
    %0 = comb.and bin %w_en, %w_mask : i1
    hw.output %memory_ext.R0_data : i32
  }
}
```

Running this through this pass produces the following (`circt-opt firrtl-snippets/circt/5681.mlir -pass-pipeline='builtin.module(hw-memory-sim{inline-mem=true})'`):

``` mlir
module {
  sv.macro.decl @RANDOM
  hw.hierpath @Bar [@Foo::@mem]
  hw.generator.schema @FIRRTLMem, "FIRRTL_Memory", ["depth", "numReadPorts", "numWritePorts", "numReadWritePorts", "readLatency", "writeLatency", "width", "maskGran", "readUnderWrite", "writeUnderWrite", "writeClockIDs", "initFilename", "initIsBinary", "initIsInline"]
  hw.module @Foo(in %r_addr : i3, in %r_en : i1, in %r_clk : i1, out r_data : i32, in %w_addr : i3, in %w_en : i1, in %w_clk : i1, in %w_data : i32, in %w_mask : i1) {
    %memory_ext_Memory = sv.reg : !hw.inout<uarray<8xi32>> 
    %memory_ext__R0_en_d0 = sv.reg sym @_R0_en_d0 : !hw.inout<i1> 
    %memory_ext__R0_addr_d0 = sv.reg sym @_R0_addr_d0 : !hw.inout<i3> 
    sv.always posedge %r_clk {
      sv.passign %memory_ext__R0_en_d0, %r_en : i1
      sv.passign %memory_ext__R0_addr_d0, %r_addr : i3
    }
    %0 = sv.read_inout %memory_ext__R0_en_d0 : !hw.inout<i1>
    %1 = sv.read_inout %memory_ext__R0_addr_d0 : !hw.inout<i3>
    %2 = sv.array_index_inout %memory_ext_Memory[%1] : !hw.inout<uarray<8xi32>>, i3
    %3 = sv.read_inout %2 : !hw.inout<i32>
    %x_i32 = sv.constantX : i32
    %4 = comb.mux %0, %3, %x_i32 : i32
    %true = hw.constant true
    sv.always posedge %w_clk {
      %6 = comb.and %5, %true : i1
      sv.if %6 {
        %7 = sv.array_index_inout %memory_ext_Memory[%w_addr] : !hw.inout<uarray<8xi32>>, i3
        %c0_i32 = hw.constant 0 : i32
        sv.passign %7, %w_data : i32
      }
    }
    sv.ifdef  "ENABLE_INITIAL_MEM_" {
      sv.ifdef  "RANDOMIZE_REG_INIT" {
        %memory_ext__RANDOM = sv.reg sym @_RANDOM : !hw.inout<i32> 
      }
      %memory_ext__RANDOM_MEM = sv.reg : !hw.inout<i32> 
      sv.initial {
        sv.verbatim "`INIT_RANDOM_PROLOG_"
        sv.ifdef.procedural  "RANDOMIZE_MEM_INIT" {
          %c0_i4 = hw.constant 0 : i4
          %c-8_i4 = hw.constant -8 : i4
          %c1_i4 = hw.constant 1 : i4
          sv.for %i = %c0_i4 to %c-8_i4 step %c1_i4 : i4 {
            %c0_i6 = hw.constant 0 : i6
            %c-32_i6 = hw.constant -32 : i6
            %c-32_i6_0 = hw.constant -32 : i6
            sv.for %j = %c0_i6 to %c-32_i6 step %c-32_i6_0 : i6 {
              %RANDOM = sv.macro.ref.se @RANDOM() : () -> i32
              %9 = sv.indexed_part_select_inout %memory_ext__RANDOM_MEM[%j : 32] : !hw.inout<i32>, i6
              sv.bpassign %9, %RANDOM : i32
            }
            %6 = comb.extract %i from 0 : (i4) -> i3
            %7 = sv.array_index_inout %memory_ext_Memory[%6] : !hw.inout<uarray<8xi32>>, i3
            %8 = sv.read_inout %memory_ext__RANDOM_MEM : !hw.inout<i32>
            sv.bpassign %7, %8 : i32
          }
        }
        sv.ifdef.procedural  "RANDOMIZE_REG_INIT" {
          sv.verbatim "{{0}} = {`RANDOM};" {symbols = [#hw.innerNameRef<@memory_8x32::@_RANDOM>]}
          sv.verbatim "{{0}} = {{1}}[0];" {symbols = [#hw.innerNameRef<@memory_8x32::@_R0_en_d0>, #hw.innerNameRef<@memory_8x32::@_RANDOM>]}
          sv.verbatim "{{0}} = {{1}}[3:1];" {symbols = [#hw.innerNameRef<@memory_8x32::@_R0_addr_d0>, #hw.innerNameRef<@memory_8x32::@_RANDOM>]}
        }
      }
    }
    %5 = comb.and bin %w_en, %w_mask : i1
    hw.output %4 : i32
  }
}
```

This is not currently caught with a verifier (it should be), but you will notice that `@mem` no longer exists and the hierarhical path op is now invalid.

Problematically, there is really nothing which is entirely correct here except exiting. The inner symbol could move to the underlying memory _register_. However, that may not be pedantically correct.

Reply:
Which approach would be preferrable then? not inline if there are inner symbols, or move them to the underlying register?

I ready the inner symbols rationale document, but I don't really have an idea when and where inner symbols are created, and whether this would be a significant limitation.

Reply:
It would be best to not inline in that case. That shouldn't affect your use case?

Reply:
I've added a check for an inner symbol on the memory module instance. If there's an inner symbol, it does not inline that instance, and makes sure the wrapper module does not get deleted.

I don't think it will affect my use case.